File: src/main/java/org/strassburger/lifestealz/commands/ReviveCommand.java
package org.strassburger.lifestealz.commands;

import net.kyori.adventure.text.Component;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.strassburger.lifestealz.LifeStealZ;
import org.strassburger.lifestealz.util.MessageUtils;
import org.strassburger.lifestealz.util.WebHookManager;
import org.strassburger.lifestealz.util.WhitelistManager;
import org.strassburger.lifestealz.util.commands.CommandUtils;
import org.strassburger.lifestealz.util.storage.PlayerData;

import java.util.List;

public class ReviveCommand implements CommandExecutor, TabCompleter {
    private final LifeStealZ plugin;

    public ReviveCommand(LifeStealZ plugin) {
        this.plugin = plugin;
    }

    private static final String BYPASS_OPTION = "bypass";
    WhitelistManager whitelistManager = new WhitelistManager();

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, String[] args) {
        if (!whitelistManager.isWorldWhitelisted(sender)) return false;

        if (args.length < 1) {
            throwUsageError(sender);
            return false;
        }

        String targetPlayersArg = args[0];

        String bypassOption = args.length > 1 ? args[1] : null;

        return targetPlayersArg.equals("*") ?
                handleReviveAll(sender, bypassOption) :
                handleReviveOne(sender, targetPlayersArg, bypassOption);
    }

    private boolean handleReviveAll(CommandSender sender, String bypassOption) {
        try {
            int revivedPlayers = plugin.getStorage().reviveAllPlayers(
                    plugin.getConfig().getInt("minHearts"),
                    plugin.getConfig().getInt("reviveHearts"),
                    plugin.getConfig().getInt("maxRevives"),
                    BYPASS_OPTION.equals(bypassOption) && sender.hasPermission("lifestealz.bypassrevivelimit")
            );
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "massReviveSuccess",
                    "&7You successfully revived &c%amount% &7player(s)!",
                    new MessageUtils.Replaceable("%amount%", Integer.toString(revivedPlayers))
            ));
            return true;
        } catch (Exception e) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "massReviveError",
                    "&cAn error occurred while reviving all players: %error%",
                    new MessageUtils.Replaceable("%error%", e.getMessage())
            ));
            plugin.getLogger().severe("An error occurred while reviving all players: " + e.getMessage());
            return false;
        }
    }

    private boolean handleReviveOne(CommandSender sender, String targetPlayerName, String bypassOption) {
        OfflinePlayer targetPlayer = plugin.getServer().getOfflinePlayer(targetPlayerName);
        PlayerData playerData = plugin.getStorage().load(targetPlayer.getUniqueId());

        if (playerData == null) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "messages.noPlayerData", "&cThis player has not played on this server yet!"));
            return false;
        }

        playerData = plugin.getStorage().load(targetPlayer.getUniqueId());

        // Check if the player has reached the revive limit or has the bypass permission
        if (!canRevive(sender, playerData, bypassOption)) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "reviveMaxReached",
                    "&cThis player has already been revived %amount% times!",
                    new MessageUtils.Replaceable("%amount%", Integer.toString(playerData.getHasbeenRevived()))));
            return false;
        }

        // Check if the player is eliminated
        if (!isEligibleForRevive(sender, playerData)) {
            sender.sendMessage(MessageUtils.getAndFormatMsg(false, "onlyReviveElimPlayers","&cYou can only revive eliminated players!"));
            return false;
        }

        revivePlayer(sender, targetPlayer.getName(), playerData);

        sender.sendMessage(MessageUtils.getAndFormatMsg(true, "reviveSuccess",
                "&7You successfully revived &c%player%&7!",
                new MessageUtils.Replaceable("%player%", targetPlayer.getName())));

        return true;
    }

    /**
     * Checks if the player can be revived
     * @param sender The command sender
     * @param playerData The player data of the player to be revived
     * @param bypassOption The bypass option
     * @return True if the player can be revived, false otherwise
     */
    private boolean canRevive(CommandSender sender, PlayerData playerData, String bypassOption) {
        int maxRevives = plugin.getConfig().getInt("maxRevives");
        boolean hasBypassPermission = sender.hasPermission("lifestealz.bypassrevivelimit");

        return maxRevives == -1 || playerData.getHasbeenRevived() < maxRevives ||
                (BYPASS_OPTION.equals(bypassOption) && hasBypassPermission);
    }

    /**
     * Checks if the player is eligible for a revive (e.g. has been eliminated)
     * @param sender The command sender
     * @param playerData The player data of the player to be revived
     * @return True if the player is eligible for a revive, false otherwise
     */
    private boolean isEligibleForRevive(CommandSender sender, PlayerData playerData) {
        int minHearts = plugin.getConfig().getInt("minHearts");

        return !(playerData.getMaxHealth() > minHearts * 2);
    }

    /**
     * Revives the player
     * @param sender The command sender
     * @param targetPlayerName The name of the player to be revived
     * @param playerData The player data of the player to be revived
     */
    private void revivePlayer(CommandSender sender, String targetPlayerName, PlayerData playerData) {
        playerData.setMaxHealth(plugin.getConfig().getDouble("reviveHearts") * 2);
        playerData.setHasbeenRevived(playerData.getHasbeenRevived() + 1);
        plugin.getStorage().save(playerData);

        plugin.getWebHookManager().sendWebhookMessage(WebHookManager.WebHookType.REVIVE, targetPlayerName, sender.getName());
    }

    /**
     * Throws a usage error message
     * @param sender The command sender
     */
    private void throwUsageError(CommandSender sender) {
        Component usageMessage = MessageUtils.getAndFormatMsg(false, "usageError",
                "&cUsage: %usage%", new MessageUtils.Replaceable("%usage%", "/revive <player>"));
        sender.sendMessage(usageMessage);
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias, String[] args) {
        if (args.length == 1) {
            return CommandUtils.getPlayersTabCompletion(true, plugin);
        }
        if (args.length == 2 && sender.hasPermission("lifestealz.bypassrevivelimit")) {
            return List.of(BYPASS_OPTION);
        }
        return null;
    }
}

File: src/main/java/org/strassburger/lifestealz/commands/ReviveCommand.kt
package org.strassburger.lifestealz.commands

import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.strassburger.lifestealz.Lifestealz
import org.strassburger.lifestealz.util.ManagePlayerdata
import org.strassburger.lifestealz.util.Replaceable

class ReviveCommand : CommandExecutor {
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>?): Boolean {

        val targetPlayerName = args?.getOrNull(0)

        val bypassOption = args?.getOrNull(1)

        if (targetPlayerName == null) {
            throwUsageError(sender)
            return false
        }

        val targetPlayer = Bukkit.getOfflinePlayer(targetPlayerName)

        if (!ManagePlayerdata().checkForPlayer(targetPlayer.uniqueId.toString())) {
            sender.sendMessage(Lifestealz.getAndFormatMsg(false, "messages.noPlayerData", "&cThis player has not played on this server yet!"))
            return false
        }

        val playerdata = ManagePlayerdata().getPlayerData(name = targetPlayerName, uuid = targetPlayer.uniqueId.toString())

        val maxRevives = Lifestealz.instance.config.getInt("maxRevives")

        if (maxRevives != -1 && playerdata.hasbeenRevived >= maxRevives && (bypassOption == null || bypassOption != "bypass" || !sender.hasPermission("lifestealz.bypassrevivelimit"))) {
            sender.sendMessage(Lifestealz.getAndFormatMsg(false, "messages.reviveMaxReached", "&cThis player has already been revived %amount% times!", Replaceable("%amount%", playerdata.hasbeenRevived.toString())))
            return false
        }

        if (playerdata.maxhp > 0.0) {
            sender.sendMessage(Lifestealz.getAndFormatMsg(false, "messages.onlyReviveElimPlayers","&cYou can only revive eliminated players!"))
            return false
        }

        ManagePlayerdata().manageOfflineHearts(name = targetPlayerName, uuid = targetPlayer.uniqueId.toString(), amount = 2.0, direction = "set")
        ManagePlayerdata().addRevive(name = targetPlayerName, uuid = targetPlayer.uniqueId.toString())

        sender.sendMessage(Lifestealz.getAndFormatMsg(true, "messages.reviveSuccess", "&7You successfully revived &c%player%&7!", Replaceable("%player%", targetPlayerName)))

        return false
    }

    fun throwUsageError(sender: CommandSender) {
        val usageMessage = Lifestealz.getAndFormatMsg(false, "messages.usageError", "&cUsage: %usage%", Replaceable("%usage%", "/revive <player>"))
        sender.sendMessage(usageMessage)
    }
}